src/Uart/Uart.c

```diff
@@ -0,0 +1,427 @@
+        *errCode = UART_ERROR_CODE_TX_FIFO_FULL;
+        return 0;
+    }
+    ByteFifo_push(uart->txFifo, data);
+    rtems_interrupt_raise(irqNumber(uart->id));
+    while((timeoutLimit == 0) || timeout-- > 0) {
+        if(ByteFifo_isEmpty(uart->txFifo) == 1) {
+            return 1;
+        }
+    }
+    ByteFifo_clear(uart->txFifo);
+    *errCode = UART_ERROR_CODE_TIMEOUT;
+    return 0;
+}
+
+bool
+Uart_read(Uart* const uart,
+          uint8_t* data,
+          uint32_t const timeoutLimit,
+          int* const errCode)
+{
+    uint8_t byte = '\0';
+    uint32_t timeout = timeoutLimit;
+    rtems_interrupt_raise(irqNumber(uart->id));
+    do {
+        if(errorFlags[uart->id].hasRxFifoFullErrorOccurred == 1) {
+            *errCode = UART_ERROR_CODE_RX_FIFO_FULL;
+            return 0;
+        }
+        if(ByteFifo_isEmpty(uart->rxFifo) == 0) {
+            ByteFifo_pull(uart->rxFifo, &byte);
+            *data = byte;
+            return 1;
+        }
+    } while((timeoutLimit == 0) || timeout-- > 0);
+    ByteFifo_clear(uart->rxFifo);
+    *errCode = UART_ERROR_CODE_TIMEOUT;
+    return 0;
+}
+
+void
+Uart_writeAsync(Uart* const uart,
+                ByteFifo* const fifo,
+                const Uart_TxHandler handler)
+{
+    rtems_interrupt_vector_disable(irqNumber(uart->id));
+    uart->txFifo = fifo;
+    uart->txHandler = handler;
+    rtems_interrupt_vector_enable(irqNumber(uart->id));
+    rtems_interrupt_raise(irqNumber(uart->id));
+}
+
+void
+Uart_readAsync(Uart* const uart,
+               ByteFifo* const fifo,
+               const Uart_RxHandler handler)
+{
+    rtems_interrupt_vector_disable(irqNumber(uart->id));
+    uart->rxFifo = fifo;
+    uart->rxHandler = handler;
+    rtems_interrupt_vector_enable(irqNumber(uart->id));
+    rtems_interrupt_raise(irqNumber(uart->id));
+}
+
+bool
+Uart_isTxEmpty(const Uart* const uart)
+{
+    return ByteFifo_isEmpty(uart->txFifo);
+}
+
+bool
+Uart_isRxEmpty(const Uart* const uart)
+{
+    return ByteFifo_isEmpty(uart->rxFifo);
+}
+
+uint32_t
+Uart_getTxFifoCount(Uart* const uart)
+{
+    return ByteFifo_getCount(uart->txFifo);
+}
+
+uint32_t
+Uart_getRxFifoCount(Uart* const uart)
+{
+    return ByteFifo_getCount(uart->rxFifo);
+}
+
+void
+Uart_handleInterrupt(void* arg)
+{
+    uint8_t buf = '\0';
+    Uart* uart = (Uart*)arg;
+    if(Uart_getLinkErrors(uart->reg->status, &(errorFlags[uart->id])) != 0) {
+        uart->errorHandler.callback(uart->errorHandler.arg);
+        return;
+    }
+    if((uart->reg->control & UART_CONTROL_RE) != 0
+       && (uart->reg->status & UART_STATUS_DR) != 0) {
+        buf = uart->reg->data;
+        if(ByteFifo_isFull(uart->rxFifo) == 1) {
+            errorFlags[uart->id].hasRxFifoFullErrorOccurred = 1;
+            return;
+        }
+        if(buf == uart->rxHandler.targetCharacter) {
+            uart->rxHandler.characterCallback(uart->rxHandler.characterArg);
+        }
+        if(sentBytesCounter[uart->id]-- == 0) {
```

Comment:
Changing a value in an if is generally a bad pattern.

Reply:
done